service/worker/batcher/workflow.go

```
@@ -20,7 +20,328 @@
+				}
+			case <-ctx.Done():
+				return nil
+			}
+		}
+
+		hbd.CurrentPage++
+		hbd.PageToken = resp.NextPageToken
+		hbd.SuccessCount += atomic.LoadInt32(&succCount)
+		hbd.ErrorCount += atomic.LoadInt32(&errCount)
+		activity.RecordHeartbeat(ctx, hbd)
+
+		if len(hbd.PageToken) == 0 {
+			break
+		}
+	}
+
+	return nil
+}
+
+func startTaskProcessor(
+	ctx context.Context, batchParams BatchParams, taskCh chan taskDetail,
+	limiter *rate.Limiter, doneCount, errCount *int32) {
+
+	for {
+		select {
+		case task := <-taskCh:
+			if isDone(ctx) {
+				return
+			}
+			var err error
+
+			switch batchParams.BatchType {
+			case BatchTypeTerminate:
+				err = processOneTerminateTask(ctx, limiter, task, batchParams)
+			}
+			if err != nil {
+				getActivityLogger(ctx).Error("Failed to process task", tag.Error(err))
+				// put back to the channel if less than attemptsOnError
+				task.attempts++
+				if task.attempts >= batchParams.AttemptsOnRetryableError {
+					atomic.AddInt32(errCount, 1)
+				} else {
+					taskCh <- task
+				}
+			} else {
+				atomic.AddInt32(doneCount, 1)
+			}
+		}
+	}
+}
+
+func processOneTerminateTask(ctx context.Context, limiter *rate.Limiter, task taskDetail, param BatchParams) error {
+	batcher := ctx.Value(batcherContextKey).(*Batcher)
+
+	wfs := []shared.WorkflowExecution{task.execution}
+	for len(wfs) > 0 {
+		wf := wfs[0]
+
+		err := limiter.Wait(ctx)
+		if err != nil {
+			getActivityLogger(ctx).Error("Failed to wait for rateLimiter", tag.Error(err))
+		}
+
+		newCtx, cancel := context.WithDeadline(ctx, time.Now().Add(rpcTimeout))
+		err = batcher.svcClient.TerminateWorkflowExecution(newCtx, &shared.TerminateWorkflowExecutionRequest{
+			Domain:            common.StringPtr(param.DomainName),
+			WorkflowExecution: &wf,
+			Reason:            common.StringPtr(param.Reason),
+			Identity:          common.StringPtr(batchWFTypeName),
+		})
+		cancel()
+		if err != nil {
+			_, ok := err.(*shared.EntityNotExistsError)
+			if !ok {
```

Shouldn't we return the error in either case? If parent terminate fails even for a retryable error we shouldnt try to terminate children.. We should try to terminate parent first..

---

That's because we need to continue on terminating all childWFs in this case. Do you have any idea to implement if we returning early here?